### PR TITLE
UefiCpuPkg: fix issue when SMM profile is enabled

### PR DESCRIPTION
**Pull request by td36**
# Description

This commit is to fix smm code assert issue when SMM Profile is enabled.

When SMM Profile is enabled, the function InitProtectedMemRange() retrives MMIO ranges from GCD and store the MMIO ranges in the mProtectionMemRange. When ReadyToLock, the function InitPaging() modifies the page table based on the mProtectionMemRange. If the MMIO ranges in mProtectionMemRange is not 4k aligned, code will assert when modifying page table.

In this commit, we skip the MMIO ranges that BaseAddress and Length are not 4k aligned when creating mProtectionMemRange. This will only cause each access to the skipped MMIO range to be logged. In current failure case on QEMU and QSP SimicsOpenBoard, the skipped MMIO range is [0xFED00000, 0xFED00400] for HPET. Considering that the probability of HPET MMIO range being accessed is very small in SMM, the solution in this commit is acceptable and simple.

## How This Was Tested
Tested on QEMU with OvmfPkg and QSP SimicsOpenBoardPkg with PcdCpuSmmProfileEnable set to TRUE

### CODE REVIEW

**Comment by td36** [comment]
hi @niruiyu @jiaxinwu @kraxel @rahul1-kumar , could you help review the PR?

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c [hunk 1]
```diff
@@ -438,7 +438,14 @@ InitProtectedMemRange (
          &MemorySpaceMap
          );
   for (Index = 0; Index < NumberOfDescriptors; Index++) {
-    if (MemorySpaceMap[Index].GcdMemoryType == EfiGcdMemoryTypeMemoryMappedIo) {
+    if ((MemorySpaceMap[Index].GcdMemoryType == EfiGcdMemoryTypeMemoryMappedIo) &&
+        ((MemorySpaceMap[Index].BaseAddress & (SIZE_4KB - 1)) == 0) &&
+        ((MemorySpaceMap[Index].Length & (SIZE_4KB - 1)) == 0))
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c:443]: can you use IS_ALIGNED() macro to check BaseAddress?
Can you simply use "%" to check Length?

Comment (niruiyu) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c:443]: 2nd comment: can you put warning here?

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c [hunk 2]
```diff
@@ -490,6 +497,18 @@ InitProtectedMemRange (
         continue;
       }
 
+      if (((MemorySpaceMap[Index].BaseAddress & (SIZE_4KB - 1)) != 0) ||
+          ((MemorySpaceMap[Index].Length & (SIZE_4KB - 1)) != 0))
+      {
```
Comment (niruiyu) [inline UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c:502]: I agree that if-check is needed here. But do you think that moving warning to above is better?

**Review by niruiyu — CHANGES_REQUESTED** [review]

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c (hunk 1 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c]: Thanks for the comments. I used ADDRESS_IS_ALIGNED for the BaseAddress and "%" for Length. Also put warning here in advance.

File: UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c (hunk 2 above)
Comment (td36) [reply UefiCpuPkg/PiSmmCpuDxeSmm/SmmProfile.c]: Moved the warning to above

**Review by niruiyu — APPROVED** [review]

**Pull request merged**